engines/e_pkcs11_eng.c

```diff
@@ -337,36 +350,35 @@ static OSSL_STORE_LOADER_CTX* pkcs11_store_open(
     PKCS11_CTX *pkcs11_ctx;
     OSSL_STORE_LOADER_CTX *store_ctx = NULL;
     char *type = NULL, *object = NULL;
+    int objecttype;
 
     store_ctx = OSSL_STORE_LOADER_CTX_new();
 
     e = (ENGINE *) OSSL_STORE_LOADER_get0_engine(loader);
     if (e == NULL)
         return NULL;
 
+    store_ctx->engine = e;
     uri += 7;
 
+    pkcs11_ctx = ENGINE_get_ex_data(e, pkcs11_idx);
+    if (pkcs11_ctx == NULL)
+        return NULL;
+
     pkcs11_parse_uri(uri,"type=", &type);
     pkcs11_parse_uri(uri,"object=", &object);
 
-    if (object != NULL && type != NULL) {
-        pkcs11_ctx = ENGINE_get_ex_data(e, pkcs11_idx);
-        if (pkcs11_ctx == NULL)
-            return NULL;
-        if (strncmp(type, "cert", 4) == 0 && !pkcs11_get_cert(
-            store_ctx, pkcs11_ctx, object))
-            return NULL;
-        if (strncmp(type, "public", 6) == 0 && !pkcs11_get_key(
-            store_ctx, pkcs11_ctx, object))
-            return NULL;
-    }
-
-    if (strncmp(uri, "list=all", 8) == 0) {
-        pkcs11_ctx = ENGINE_get_ex_data(e, pkcs11_idx);
-        if (pkcs11_ctx == NULL)
-            return NULL;
-        if (!pkcs11_get_ids(store_ctx, pkcs11_ctx))
-            return NULL;
+    if (type != NULL) {
+        objecttype = pkcs11_get_type(type);
+        switch (objecttype) {
+        case OTYPE_CERT:
+            if (!pkcs11_get_cert(store_ctx, pkcs11_ctx, object))
+                return NULL;
+            break;
+        case OTYPE_PUBLIC:
+            if (!pkcs11_get_key(store_ctx, pkcs11_ctx, object))
+                return NULL;
+        }     
```

Comment:
This still seems unnecessary.  `OSSL_STORE` is about discovery, so rather than deciding through the URI what the type should be, it should be enough to give the object label, and then the `OSSL_STORE` load function should be able to check what object it got and return an appropriate `OSSL_STORE_INFO` for each of them.

Reply:
Maybe there's something I don't understand.
In the file URI Scheme object is unique but in PKCS 11 URI Scheme (https://tools.ietf.org/html/rfc7512) it is not.
The same label (object) identifies, for example, certificate and public key.
If the user needs to download the certificate or public key, how can he do without indicating the type?

Reply:
A PKCS#11 URI is basically a search term. A STORE for a given URI should contain all the objects identified by the corresponding search. The URI may, or may not, include an explicit `type` specification. If it doesn't, you're likely to find both a key and a certificate in the store.

It's similar for a file-based STORE. Often you can find certificates and keys stored in the same file. I don't think there's anything special and different about the PKCS#11 STORE here, is there?

Reply:
`OSSL_STORE` has a couple of functions to help filtering, one of them is `OSSL_STORE_expect`, and there is a corresponding loader callback that can be set with `OSSL_STORE_LOADER_set_expect`.  But if not using that, the users are expected to filter on their own, and they can use `OSSL_STORE_INFO_get_type` to figure out what kind of object they got.

The point is that the user should have the choice of getting *everything* that's associated with an identity.  Forcing them to only look for one type of object isn't very helpful with that in mind...

Regarding the file-based `OSSL_STORE`, a PEM file can contain several objects.  You may actually often find PEM files that contain a certificate and its corresponding private key.  If you use `openssl storeutl` on such a PEM file, you get this:

```
# openssl storeutl -noout /etc/ejabberd/ejabberd.pem
0: Pkey
1: Certificate
Total found: 2
```

The PKCS#11 engine should be capable of delivering the same kind of output for something like `pkcs11:object=foo` if that one has several objects associated with it.

Reply:
Of course, if the user *chooses* to add additional URI attributes to only allow specific object types, that's absolutely fine.  However, that choice shouldn't be mandatory.